File: admin-web/src/lib/mock/billing.ts
/**
 * 账单中心假数据（SaaS 增强项，epay 无直接对应）。
 * 平台财务对账视图：按月归集收入（手续费利润）与支出（结算打款 / 代付 / 分账 / 退款），
 * 生成可对账、可导出的月度账单。数据口径与订单/结算/分账/付款各模块一致。
 */

/** 账单收支分项 */
export interface BillItem {
  label: string
  amount: number
  kind: 'income' | 'expense'
}

/** 月度账单 */
export interface MonthlyBill {
  period: string // 账期 YYYY-MM
  opening: number // 期初平台留存
  incomes: BillItem[] // 收入分项
  expenses: BillItem[] // 支出分项
  status: 0 | 1 // 0=待对账 1=已对账
}

function round(n: number) {
  return +n.toFixed(2)
}

// 生成近 6 个月账单（确定性）
const months = ['2026-07', '2026-06', '2026-05', '2026-04', '2026-03', '2026-02']

export const bills: MonthlyBill[] = months.map((period, i) => {
  const base = 100000 - i * 8000
  const feeIncome = round(base * 0.018) // 手续费利润
  const channelCost = round(base * 0.006) // 通道成本支出
  const settleOut = round(base * 0.9) // 结算打款
  const transferOut = round(base * 0.12) // 代付打款
  const psOut = round(base * 0.05) // 分账支出
  const refundOut = round(base * 0.02) // 退款支出
  return {
    period,
    opening: round(50000 + i * 1200),
    incomes: [
      { label: '订单手续费利润', amount: feeIncome, kind: 'income' as const },
      { label: '会员套餐售卖', amount: round(base * 0.004), kind: 'income' as const },
      { label: '其他收入', amount: round(base * 0.001), kind: 'income' as const },
    ],
    expenses: [
      { label: '商户结算打款', amount: settleOut, kind: 'expense' as const },
      { label: '代付转账', amount: transferOut, kind: 'expense' as const },
      { label: '分账支出', amount: psOut, kind: 'expense' as const },
      { label: '订单退款', amount: refundOut, kind: 'expense' as const },
      { label: '通道成本', amount: channelCost, kind: 'expense' as const },
    ],
    status: (i === 0 ? 0 : 1) as 0 | 1,
  }
})

/** 账单派生汇总 */
export function billSummary(b: MonthlyBill) {
  const income = b.incomes.reduce((a, x) => a + x.amount, 0)
  const expense = b.expenses.reduce((a, x) => a + x.amount, 0)
  const net = income - expense
  return {
    income: round(income),
    expense: round(expense),
    net: round(net),
    closing: round(b.opening + net),
  }
}

/** 账单状态字典 */
export const billStatus: Record<number, { text: string; variant: 'warning' | 'success' }> = {
  0: { text: '待对账', variant: 'warning' },
  1: { text: '已对账', variant: 'success' },
}

File: 0538pay-api/internal/service/billing.go
package service

import (
	"time"

	"github.com/epvia/api/internal/dto"
	"github.com/epvia/api/internal/repository"
)

// BillingService 平台月度对账账单聚合（我方独有细化项，epay 无）。
// 按账期从 订单/结算/代付/分账/退款 五张既有真实表归集平台资金收支，生成可对账的月度账单。
type BillingService struct {
	repo *repository.BillingRepo
}

func NewBillingService(repo *repository.BillingRepo) *BillingService {
	return &BillingService{repo: repo}
}

// defaultBillMonths 账单中心默认回溯的账期数（含当月）。
const defaultBillMonths = 6

// Billing 生成近 months 期（含当月）的平台月度账单，倒序返回（最新账期在前）。
// months<=0 时用默认值 defaultBillMonths。
func (s *BillingService) Billing(months int) (*dto.BillingResult, error) {
	if months <= 0 {
		months = defaultBillMonths
	}
	now := time.Now()
	// 当月月初（本地时区）。
	curMonthStart := time.Date(now.Year(), now.Month(), 1, 0, 0, 0, 0, now.Location())

	res := &dto.BillingResult{}
	for i := 0; i < months; i++ {
		start := curMonthStart.AddDate(0, -i, 0)
		end := start.AddDate(0, 1, 0)
		bill, err := s.buildMonth(start, end)
		if err != nil {
			return nil, err
		}
		// 当月(i==0)标进行中，历史月份标已归档。
		if i == 0 {
			bill.Status = 0
		} else {
			bill.Status = 1
		}
		res.Bills = append(res.Bills, *bill)
	}
	return res, nil
}

// buildMonth 归集单个账期 [start,end) 的收支。
func (s *BillingService) buildMonth(start, end time.Time) (*dto.MonthlyBill, error) {
	profit, err := s.repo.SumOrderProfit(start, end)
	if err != nil {
		return nil, err
	}
	settleOut, err := s.repo.SumSettleReal(start, end)
	if err != nil {
		return nil, err
	}
	transferOut, err := s.repo.SumTransferMoney(start, end)
	if err != nil {
		return nil, err
	}
	profitOut, err := s.repo.SumProfitMoney(start, end)
	if err != nil {
		return nil, err
	}
	refundOut, err := s.repo.SumRefundMoney(start, end)
	if err != nil {
		return nil, err
	}
	orders, err := s.repo.CountPaidOrders(start, end)
	if err != nil {
		return nil, err
	}

	incomes := []dto.BillItem{
		{Label: "订单手续费利润", Amount: profit.StringFixed(2), Kind: "income"},
	}
	expenses := []dto.BillItem{
		{Label: "商户结算打款", Amount: settleOut.StringFixed(2), Kind: "expense"},
		{Label: "代付转账", Amount: transferOut.StringFixed(2), Kind: "expense"},
		{Label: "分账划拨", Amount: profitOut.StringFixed(2), Kind: "expense"},
		{Label: "订单退款", Amount: refundOut.StringFixed(2), Kind: "expense"},
	}

	incomeSum := profit
	expenseSum := settleOut.Add(transferOut).Add(profitOut).Add(refundOut)
	net := incomeSum.Sub(expenseSum)

	return &dto.MonthlyBill{
		Period:   start.Format("2006-01"),
		Orders:   orders,
		Incomes:  incomes,
		Expenses: expenses,
		Income:   incomeSum.StringFixed(2),
		Expense:  expenseSum.StringFixed(2),
		Net:      net.StringFixed(2),
	}, nil
}
